### PR TITLE
fix AD symmetricEigenvaluesEigenvectors dead loop

### PR DESCRIPTION
**Pull request by hugary1995**
Kill the loop when it is not converging. Using an off-diagonal shift if diagonal entries are all zero.

refs #13652

### CODE REVIEW

File: framework/src/utils/RankTwoTensor.C [hunk 1]
```diff
@@ -1067,7 +1067,14 @@ RankTwoTensorTempl<DualReal>::symmetricEigenvaluesEigenvectors(
     do
     {
       iter++;
+      if (iter > 100)
+        mooseError("eigenvalue decomposition does not converge.");
       DualReal shift = D(m, m);
+      if (MooseUtils::absoluteFuzzyEqual(shift.value(), 0.0))
+      {
+        shift = std::max(D(m, m), D(m - 1, m - 1));
+        shift = std::max(D(m - 1, m), shift);
```
Comment (dschwen) [inline framework/src/utils/RankTwoTensor.C]: just do nested max

File: framework/src/utils/RankTwoTensor.C [hunk 2]
```diff
@@ -1067,7 +1067,14 @@ RankTwoTensorTempl<DualReal>::symmetricEigenvaluesEigenvectors(
     do
     {
       iter++;
+      if (iter > 100)
```
Comment (dschwen) [inline framework/src/utils/RankTwoTensor.C]: please make this a `mooseException` instead

**Review by dschwen — CHANGES_REQUESTED** [review]

File: framework/src/utils/RankTwoTensor.C (hunk 2 above)
Comment (dschwen) [reply framework/src/utils/RankTwoTensor.C]: I mean the `mooseError` below, of course

**Comment by dschwen** [comment]
Can you please add a test case in the unit tests where the diagonal is zero?

**Comment by dschwen** [comment]
Funny enough I may have run into this issue already! I had some random models freeze up after converting them to AD...

**Comment by hugary1995** [comment]
Sure, there are also a few other test cases missing for this. I'll add them.

**Comment by hugary1995** [comment]
I've tried my best, there are simply some cases that don't work. The two failing cases have correct values but incorrect derivatives. I cannot say that the chance for those two failing cases to happen is negligible, but all my use cases work fine with the current version. Perhaps you can test this on your previously frozen application?

=====================
Good: three non-zero eigenvalues
        39.8272 {   13  }       -4.30189 {   -9  }        8.41378 {    2  }
       -4.30189 {   -9  }        20.5534 {  -22  }         4.3064 {    1  }
        8.41378 {    2  }         4.3064 {    1  }       -2.38058 {   70  }
=====================
Good: two non-zero eigenvalues
        40.0252 {   13  }       -4.10753 {   -9  }        7.45827 {    2  }
       -4.10753 {   -9  }        20.7442 {  -22  }        3.36843 {    1  }
        7.45827 {    2  }        3.36843 {    1  }        2.23063 {   70  }
=====================
Bad: One non-zero eigenvalues
        39.7135 {   13  }       -6.58244 {-57.1785  }        6.89027 {-44.0261  }
       -6.58244 {-57.1785  }        1.09103 {41.1494  }       -1.14205 {-1.88723  }
        6.89027 {-44.0261  }       -1.14205 {-1.88723  }        1.19546 {6.85064  }
=====================
Good: Diagonal matrix
             42 {   13  }   -5.85031e-16 {   -9  }              0 {    2  }
   -5.85031e-16 {   -9  }             21 {  -22  }              0 {    1  }
              0 {    2  }              0 {    1  }              0 {   70  }
=====================
Good: Off-diagonal matrix
   -5.41789e-14 {   13  }              5 {   -9  }              0 {    2  }
              5 {   -9  }    5.24025e-14 {  -22  }              0 {    1  }
              0 {    2  }              0 {    1  }              0 {   70  }
=====================
Bad: zero matrix
              0 { -4.5  }              0 {   -9  }              0 {    0  }
              0 {   -9  }              0 { -4.5  }              0 {    0  }
              0 {    0  }              0 {    0  }              0 {   70  }

**Comment by dschwen** [comment]
Are we even sure the eigen decomposition is differentiable in those cases?

**Comment by hugary1995** [comment]
That's a good question, in general I think when two eigenvalues are the same, the projection is non-differentiable. You can see that in the hand-coded version, i.e. "positiveProjectionEigenDecomposition", we have an approximation using L'Hopital's rule in case of equal eigenvalues:

```c++
T theta_ab;
if (!MooseUtils::absoluteFuzzyEqual(eigval[a], eigval[b]))
  theta_ab = 0.5 * (epos[a] - epos[b]) / (eigval[a] - eigval[b]);
else
  theta_ab = 0.25 * (d[a] + d[b]);

proj_pos += theta_ab * (Gab + Gba);
```

I thought AD can catch this case automatically but now I think I might be wrong. Actually all the failing cases have repeated eigenvalues.

**Comment by dschwen** [comment]
In https://www.win.tue.nl/analysis/reports/rana06-33.pdf the derivatives are derived only for non-defective matrices (that paper is also referenced in [this tensorflow issue]). This PR seems to be an improvement. Let me see if I can replicate my freezes.

**Pull request closed without merging**